code/assignTime.m

```diff
@@ -0,0 +1,281 @@
+firstGoodIndex = find(dataTable_original.PacketGenTime > 0,1);
+
+normedTimestamps = dataTable_original.timestamp - dataTable_original(firstGoodIndex,:).timestamp;
+normedGenTime_inSecs = (dataTable_original.PacketGenTime - dataTable_original(firstGoodIndex,:).PacketGenTime)/1000;
+
+timeDifferences = normedTimestamps - normedGenTime_inSecs;
+indices_outlierPacketGenTimes = find(abs(timeDifferences) > 1);
+
+% Identify packetGenTimes that go backwards in time; should overlap with negative PacketGenTime
+packetGenTime_diffs = diff(dataTable_original.PacketGenTime);
+diffIndices = find(packetGenTime_diffs < 0 );
+
+% Need to remove [diffIndices + 1], but may also need to remove subsequent
+% packets. Remove at most 5 adjacent packets (to prevent large un-needed
+% packet rejection driven by positive outliers)
+indices_backInTime = [];
+for iIndex = 1:length(diffIndices)
+    counter = 1;
+    while (counter < 6) && dataTable_original.PacketGenTime(diffIndices(iIndex) + counter)...
+            < dataTable_original.PacketGenTime(diffIndices(iIndex))
+        
+        indices_backInTime = [indices_backInTime (diffIndices(iIndex) + counter)];
+        counter = counter + 1;
+    end
+end
+
+% Collect all packets to remove
+packetsToRemove = unique([badDatePackets; packetIndices_NegGenTime;...
+    duplicate_firstIndex + 1; indices_outlierPacketGenTimes; indices_backInTime']);
+
+% Remove packets identified above for rejection
+packetsToKeep = setdiff(1:size(dataTable_original,1),packetsToRemove);
+dataTable = dataTable_original(packetsToKeep,:);
+
+%%
+% Identify gaps -- start with most obvious gaps, and then become more
+% refined
+
+% Change in sampling rate should start a new chunk; identify indices of
+% last packet of a chunk
+if ~isequal(length(unique(dataTable.samplerate)),1)
+    indices_changeFs = find(diff(dataTable.samplerate) ~=0);
+else
+    indices_changeFs = [];
+end
+maxFs = max(unique(dataTable.samplerate));
+
+% Timestamp difference from previous packet not 0 or 1 -- indicates gap or
+% other abnormality; these indices indicate the end of a continuous chunk
+indices_timestampFlagged = intersect(find(diff(dataTable.timestamp) ~= 0),find(diff(dataTable.timestamp) ~= 1));
+
+% Instances where dataTypeSequence doesn't iterate by 1; these indices
+% indicate the end of a continuous chunk
+indices_dataTypeSequenceFlagged = intersect(find(diff(dataTable.dataTypeSequence) ~= 1),find(diff(dataTable.dataTypeSequence) ~= -255));
+
+% Lastly, check systemTick to ensure there are no packets missing.
+% Determine delta time between systemTicks in adjacent packets.
+% Exclude first packet from those to test, because need to compare times from
+% 'previous' to 'current' packets; diff_systemTick written to index of
+% second packet associated with that calculation
+numPackets = size(dataTable,1);
+for iPacket = 2:numPackets
+    diff_systemTick(iPacket,1) = mod((dataTable.systemTick(iPacket) + (2^16)...
```

Comment:
If speed is a concern, preallocate 'diff_systemTick' to the same size as dataTable. Not a big deal.

Does systemTick go from [0,...,2^16-1] or [1,...,2^16]? I think in either case that this calculation simplifies to 
`mod((dataTable.systemTick(iPacket) - dataTable.systemTick(iPacket-1)),2^16)` 
because 2^16%2^16 = 0

Please correct me if I'm wrong. Or if there is a special case I'm missing?

Reply:
@kenlouie I believe you need the `2^16` added to the `iPacket` value because `systemTick` could have wrapped. In other words, the `iPacket` value could be smaller than the `iPacket-1` value. Does this make sense?

Reply:
Preallocating diff_systemTick with 8e1bfb29e74441410d63ea472713630ae862218e

I believe systemTick goes from 0 to 2^16 - 1, but not positive

Reply:
Yes, that makes sense. I posted an argument to this point below for completeness, but it is mostly just technicalities, doesn't change the results, and is not as intuitive for people to pickup right away. So this line should stay the same as it is more clear, and the addition of that operation to run time is puny. I'm marking this as resolved. Thanks for the conversation and answering my question about the sys tick range!


------ Boring stuff below -------

Looking up the definition of the modulus operator, the equation is: mod(x,n) = x - n * floor(x / n) where x is the value and n is the divisor. This definition is also referenced in the official matlab documentation. From the equation, and fixing n = 2^16, we can see that floor(x/n) will always give us -1 if we pass it in a negative number between [-2^16,..,0]. For example, if the sys tick is 61 for the current packet and sys tick is 62291 for the previous packet, x = -62230 and floor(-62230/2^16) = -1. Continuing in the equation, we get mod = -62230 - (2^16*-1) = -6223 + 2^16 = 3306 = (2^16-62291) + 61. I compared outcomes of the code above with and without 2^16 added and it returned the same values. I never knew how modulus handled negative numbers, so I thank you for making me check this!